### DISCUSSION TITLE
Possible off the shelf hardware option - T-RSC3

### DISCUSSION THREAD

**Discussion by netmindz** [netmindz/balboa_GL_ML_spa_control#49] [Ideas]
I have ordered myself one of these boards to try

It has the voltage regulator we need and the RS485 support so it's only really the direct application of 5v to the GPIO pin for panel select that isn't ideal

LILYGO® T-RSC3 ESP32-C3 Development Board RS232 RS485 5V Module On-Board WS2812 Wi-Fi Bluetooth 5.0 Typs C RSM232MT5V RSM485MT5V
https://a.aliexpress.com/_EI4V003

I'll report back how I get on

**Comment by netmindz** [comment]
Hardware has arrived now so hopefully test soon. 

It also has a neopixel. So that would be handy for showing status

**Comment by netmindz** [comment]
Code changes - https://github.com/netmindz/balboa_GL_ML_spa_control/pull/55

**Comment by netmindz** [comment]
The sending of commands is *NOT* working yet. I'm currently using the hardware however for ESPHome testing

**Comment by netmindz** [comment]
Have switched hardware over and even with the new improved status reading and sending the right size message sending is not working for some reason on this hardware. Not sure of the chip used, but I'm treating it as if it's a max485

**Comment by netmindz** [comment]
Looks like at least one other is having issues with broken output on this board https://github.com/Xinyuan-LilyGO/T-RSC3/issues/3

**Reply by davewatson91** [reply]
This is interesting - I wonder what could possibly be causing that - I translated the datasheet for the RS485 transceiver - not sure if there is anything of value on there for you? 
[RSM485MT5V_CN_v1.0 (1).pdf](https://github.com/netmindz/balboa_GL_ML_spa_control/files/13260534/RSM485MT5V_CN_v1.0.1.pdf)

Also came across this example - I haven't checked to see if it varies from what you're already doing.

I suspect you've seen these both already, given your link above, but figured worth sharing just in-case.

**Comment by netmindz** [comment]
I wonder if https://github.com/espressif/arduino-esp32/pull/5549 is actually included in my esphome builds

**Comment by davewatson91** [comment]
So I managed to get my tub reading & sending (although not particularly smoothly). I'm not 100% sure what I changed tbh, but it's now working 🙄 

Basically, the sending seems to get confused / locked up a bit - it latched up and broke everything, where I had to hard kill the hot tub, remove the power to the ESP, and then reconnect it after powering the spa back up. 

It seemed to be when trying to send a couple of commands in a row (i.e. changing temperature). I need to dig through the code to understand how each of the commands is being sent (i.e. difference between the HA dropdown for mode, and the mode button (my panel doesn't have a mode button). 

Aux & Pump2 behaviour is a little off as well - I have a 24x7 circulation pump, main pump & blower - Pump1 is my main pump, but Aux & Pump2 both will turn on the blower, but neither will turn it off. It also won't show the state of the blower (or the circulation pump).

**Reply by netmindz** [reply]
Great to hear you have been able to send data with your board. Wonder if mine is just broken then.

What panel do you have?

When you press buttons on the existing panel do you see different values for the FB sensor? It might change too quickly so you might need to watch the MQTT data directly. If you do, then we can use this as an easy way to capture the commands sent by your panel to tweak what is expected.

I have a dedicated circulation pump, but it doesn't run 24/7

**Reply by davewatson91** [reply]
It's an ML400 panel (like the file below), except mine has 'air' where it has 'aux'.. 

[40570-99_C.pdf](https://github.com/netmindz/balboa_GL_ML_spa_control/files/13289533/40570-99_C.pdf)

Unfortunately the FB values don't change for anything (confirmed using MQTT Explorer (which retains a history even if it does change quick). 

Changing the main jets, changes pump1 (and a few other entries), air does update status and a couple of others, but only the generic things.

**Reply by netmindz** [reply]
If you use the record-fb branch that uses an older way of handing the serial data that allows you to see the messages being sent by your existing panel, the FB values. This should help debug what is different about your setup. If you share them here I can hopefully help see what's going on and then we can tweak what is sent by our own code

**Reply by davewatson91** [reply]
Thanks! I'm away for a week, but will have a go when I get back :)

**Reply by netmindz** [reply]
Any luck @davewatson91 

I recently pushed a few optimisations that hopefully improve command sending reliability
